src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs

```diff
@@ -167,43 +178,95 @@ private static SyntaxNode GetEnclosingCodeElementNode(Document document, SyntaxT
 
         public async Task<IEnumerable<ReferenceLocationDescriptor>> FindReferenceLocationsAsync(Solution solution, DocumentId documentId, SyntaxNode syntaxNode, CancellationToken cancellationToken)
         {
-            var document = solution.GetDocument(documentId);
-            if (document == null)
+            return await FindAsync(solution, documentId, syntaxNode,
+                async progress =>
+                {
+                    var referenceTasks = progress.Locations
+                        .Where(location => location.Kind != LocationKind.MetadataFile && location.Kind != LocationKind.None)
+                        .Distinct(LocationComparer.Instance)
+                        .Select(
+                            location =>
+                                GetDescriptorOfEnclosingSymbolAsync(solution, location, cancellationToken))
+                        .ToArray();
+
+                    await Task.WhenAll(referenceTasks).ConfigureAwait(false);
+
+                    return referenceTasks.Select(task => task.Result);
+                }, onCapped: null, searchCap: 0, cancellationToken: cancellationToken).ConfigureAwait(false);
+        }
+
+        private static SyntaxNode FindEnclosingMethodNode(Document document, int position)
+        {
+            var getSyntaxRootTask = document.GetSyntaxRootAsync();
```

Comment:
GetSyntaxRootAsync.Result can be both very expensive and blocking.  This should be await'ed and should have a cancellation token passed to it.